{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "748008f6-a35b-476e-b377-ff5a8bc00352",
   "metadata": {},
   "source": [
    "# Imports for feature selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b13d97c5-e032-49ff-b5fe-89bad35f2ee5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sksurv.linear_model import CoxPHSurvivalAnalysis, CoxnetSurvivalAnalysis\n",
    "from sksurv.util import Surv\n",
    "from sklearn.pipeline import Pipeline, make_pipeline\n",
    "from sklearn.feature_selection import SelectorMixin, VarianceThreshold\n",
    "from sklearn.preprocessing import StandardScaler, FunctionTransformer, Normalizer, RobustScaler\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.compose import make_column_transformer, make_column_selector, ColumnTransformer\n",
    "from sklearn.base import BaseEstimator, TransformerMixin\n",
    "from sklearn.experimental import enable_iterative_imputer\n",
    "from sklearn.impute import IterativeImputer\n",
    "from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69c2f133-f407-4333-843b-e5b94d6cb4dd",
   "metadata": {},
   "source": [
    "# Imports from main.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d66028bd-e0d7-4345-8ff1-f82c453fd526",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from main.py\n",
    "import os\n",
    "os.chdir('/home/users/nus/e1083772/cancer-survival-ml/')\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "shuffle=1\n",
    "fold=4\n",
    "endpoint='os'\n",
    "\n",
    "scratchdir=\"/scratch/users/nus/e1083772/cancer-survival-ml/data/splits\"\n",
    "features_file=f'{scratchdir}/{shuffle}/{fold}/train_features.parquet'\n",
    "features = pd.read_parquet(features_file)\n",
    "\n",
    "valid_features_file=f'{scratchdir}/{shuffle}/{fold}/valid_features.parquet'\n",
    "valid_features = pd.read_parquet(valid_features_file)\n",
    "\n",
    "survcols = [f'{endpoint}cdy',f'cens{endpoint}']\n",
    "train_surv_file=f'{scratchdir}/{shuffle}/{fold}/train_labels.parquet'\n",
    "train_surv = pd.read_parquet(train_surv_file,columns=survcols)\n",
    "train_surv.rename(columns={f'{endpoint}cdy':'survtime',f'cens{endpoint}':'survflag'},inplace=True)\n",
    "\n",
    "train_out_features_file=f'{scratchdir}/{shuffle}/{fold}/features_processed.parquet'\n",
    "valid_out_features_file=f'{scratchdir}/{shuffle}/{fold}/valid_features_processed.parquet'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cbabd343-e1df-4f8f-86d7-5139dad5fa95",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1016, 115442)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e4ef6202-43bd-432d-8cd4-c283997b5e8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1016, 2)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_surv.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "075e5165-5227-4c4e-a5b4-471487af9ce5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(127, 115442)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "valid_features.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37ba371b-982c-452c-96c8-dca2137c04fa",
   "metadata": {},
   "source": [
    "# Adjust negative survival times"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a4cdb9c6-ecdd-4397-b252-d4d453eb0f59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21\n"
     ]
    }
   ],
   "source": [
    "event = train_surv['survflag'].values\n",
    "time = train_surv.survtime\n",
    "offset = max(0, -np.min(time)) # some OS is negative\n",
    "print(offset)\n",
    "time += offset\n",
    "y = Surv.from_arrays(event,time)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98a67482-9882-4f49-8f4d-1c26d93ab457",
   "metadata": {},
   "source": [
    "# Feature selection on RNA-Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fe63f353-8ecb-4644-9a27-b085ad6db65e",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna = features.filter(regex='Feature_exp_')\n",
    "rna_cols = rna.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6fba2c75-65dc-42dd-9f3c-d73c06a12389",
   "metadata": {},
   "outputs": [],
   "source": [
    "cna = features.filter(regex='Feature_CNA_ENSG')\n",
    "cna_cols = cna.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d01db0e0-b794-4d33-9cfa-f4f2240c9fcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "gistic = features.filter(regex='Feature_CNA_(Amp|Del)')\n",
    "gistic_cols = gistic.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "688e66e3-abd6-4a8a-afb3-1d499709c490",
   "metadata": {},
   "outputs": [],
   "source": [
    "fish = features.filter(regex='Feature_fish')\n",
    "fish_cols = fish.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ce898ab8-c2ec-4e72-bc8a-451f382b3aaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "clin = features.filter(regex='Feature_clin')\n",
    "clin_cols = clin.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a39bb1e4-2f43-4226-9e97-034a6409ef85",
   "metadata": {},
   "outputs": [],
   "source": [
    "ig = features.filter(regex='Feature_SeqWGS')\n",
    "ig_cols = ig.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ca066dfc-4ec9-460d-a73d-07f8f1bdfd96",
   "metadata": {},
   "outputs": [],
   "source": [
    "sbs = features.filter(regex='Feature_SBS')\n",
    "sbs_cols = sbs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "208f5778-0d6b-47a9-8ac7-a1dbf11348e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "apobec = features.filter(regex='APOBEC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "14ac322f-03e6-4cc8-a44e-d80aa645145f",
   "metadata": {},
   "outputs": [],
   "source": [
    "chromothripsis = features.filter(regex='chromothripsis')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6832f3e2-4db5-4a48-aa30-ba679275651e",
   "metadata": {},
   "source": [
    "# TransformerMixin functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "3f69b728-e550-4af1-a15a-c8e651acdf91",
   "metadata": {},
   "outputs": [],
   "source": [
    "class VarianceSelector(TransformerMixin, BaseEstimator):        \n",
    "    def __init__(self, **kwargs):\n",
    "        self.vt = VarianceThreshold(**kwargs)\n",
    "        self.features_out = None\n",
    "        \n",
    "    def fit(self, X, y=None, **kwargs):\n",
    "        self.vt.fit(X)\n",
    "        self.features_out = self.vt.get_feature_names_out()\n",
    "        return self\n",
    "    \n",
    "    def transform(self, X):\n",
    "        Xr = self.vt.transform(X)\n",
    "        return pd.DataFrame(Xr, columns=self.features_out, index=X.index)\n",
    "\n",
    "    def get_feature_names_out(self):\n",
    "        return self.features_out\n",
    "\n",
    "class CoxnetSelector(TransformerMixin, BaseEstimator):\n",
    "    def __init__(self,coef_threshold=0,**kwargs):\n",
    "        self.cns = CoxnetSurvivalAnalysis(**kwargs)\n",
    "        self.coef_threshold = coef_threshold\n",
    "        self.features_out = None\n",
    "        \n",
    "    def fit(self, X, y, **kwargs):\n",
    "        nan = np.isnan(X).any(axis=1).values\n",
    "        notnan = np.where(~nan)[0]\n",
    "        self.cns.fit(X.iloc[notnan,:], y[notnan])\n",
    "        _keep = np.abs(self.cns.coef_[:,-1]) > self.coef_threshold\n",
    "        features_out = np.where(_keep)[0]\n",
    "        self.features_out = X.columns[features_out]\n",
    "        return self\n",
    "    \n",
    "    def transform(self, X):\n",
    "        return X.loc[:,self.features_out]\n",
    "\n",
    "    def get_feature_names_out(self):\n",
    "        return np.array(self.features_out)\n",
    "\n",
    "class StandardTransform(TransformerMixin, BaseEstimator):\n",
    "    # scale a selection of features\n",
    "    def __init__(self, cols=None):\n",
    "        self.scaler = StandardScaler()\n",
    "        self.cols = cols # columns to scale\n",
    "\n",
    "    def fit(self, X, y=None):\n",
    "        if self.cols is None:\n",
    "            self.cols = X.columns\n",
    "        self.scaler.fit(X[self.cols])\n",
    "        self.feature_names_in = X.columns\n",
    "        return self\n",
    "\n",
    "    def transform(self, X):\n",
    "        Xt = X.astype({col: 'float' for col in self.cols})\n",
    "        Xt.loc[:,self.cols] = self.scaler.transform(Xt[self.cols])\n",
    "        return Xt\n",
    "\n",
    "    def get_feature_names_out(self):\n",
    "        return np.array(self.feature_names_in)\n",
    "\n",
    "class CorrelationSelector(TransformerMixin, BaseEstimator):\n",
    "    def __init__(self, threshold=0.95):\n",
    "        self.threshold = threshold\n",
    "        self.features_out = None\n",
    "\n",
    "    def fit(self, X, y=None):\n",
    "        assert isinstance(X, pd.DataFrame)\n",
    "        X1 = X.copy()\n",
    "        corr_matrix = X.corr()\n",
    "        col_corr = set() # correlated (deleted) columns\n",
    "        for i in range(len(corr_matrix.columns)):\n",
    "            for j in range(i):\n",
    "                if (abs(corr_matrix.iloc[i, j]) >= self.threshold) and (corr_matrix.columns[j] not in col_corr):\n",
    "                    colname = corr_matrix.columns[i]\n",
    "                    col_corr.add(colname)\n",
    "                    if colname in X1.columns:\n",
    "                        del X1[colname]\n",
    "        self.features_out = X1.columns\n",
    "        return self\n",
    "    \n",
    "    def transform(self, X):\n",
    "        return X.loc[:, self.features_out]\n",
    "    \n",
    "    def get_feature_names_out(self):\n",
    "        return self.features_out\n",
    "\n",
    "class FrequencySelector(TransformerMixin, BaseEstimator):\n",
    "    def __init__(self, minfreq=0.05, mincount=np.Inf):\n",
    "        # default is to use frequency cutoff\n",
    "        self.minfreq = minfreq\n",
    "        self.mincount = mincount\n",
    "        self.features_out = None\n",
    "\n",
    "    def fit(self, X, y=None):\n",
    "        assert isinstance(X, pd.DataFrame)\n",
    "        counts = X.sum(axis=0)\n",
    "        freqs = counts / (~X.isna()).sum(axis=0)\n",
    "        n_max = (~X.isna()).sum(axis=0).max()\n",
    "        usecols = (counts >= self.mincount) if self.mincount/n_max < self.minfreq else freqs >= self.minfreq\n",
    "        self.features_out = np.array(X.columns[usecols])\n",
    "        return self\n",
    "\n",
    "    def transform(self, X):\n",
    "        return X.loc[:, self.features_out]\n",
    "\n",
    "    def get_feature_names_out(self):\n",
    "        return self.features_out\n",
    "\n",
    "class Log1pTransform(TransformerMixin, BaseEstimator):\n",
    "    def __init__(self,):\n",
    "        self.features_out = None \n",
    "        \n",
    "    def fit(self, X, y=None):\n",
    "        self.features_out = X.columns\n",
    "        return self\n",
    "        \n",
    "    def transform(self, X):\n",
    "        logX = np.log1p(X)\n",
    "        outX = pd.DataFrame(logX, index=X.index, columns=self.features_out)\n",
    "        return outX \n",
    "    \n",
    "    def get_feature_names_out(self):\n",
    "        return self.features_out\n",
    "    \n",
    "\n",
    "class PCATransform(TransformerMixin, BaseEstimator):\n",
    "    # accepts NA values unlike normal PCA\n",
    "    def __init__(self, prefix=None, **kwargs):\n",
    "        self.prefix = prefix if prefix else 'Unnamed_'\n",
    "        self.features_out = None\n",
    "        self.pca = PCA(**kwargs) # e.g. n_components\n",
    "        \n",
    "    def fit(self, X, y=None):\n",
    "        Xfull = X.dropna()\n",
    "        self.pca.fit(Xfull)\n",
    "        self.features_out = np.array([self.prefix + str(s) for s in self.pca.get_feature_names_out()])\n",
    "        return self\n",
    "    \n",
    "    def transform(self, X):\n",
    "        Xfull = X.dropna()\n",
    "        pcX = self.pca.transform(Xfull)\n",
    "        outX = pd.DataFrame(pcX, \n",
    "                            index=Xfull.index, \n",
    "                            columns=self.features_out).reindex(X.index)\n",
    "        return outX\n",
    "    \n",
    "    def get_feature_names_out(self):\n",
    "        return self.features_out"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28b5b040-f0d8-4013-94ce-f9dbc74b2b53",
   "metadata": {},
   "source": [
    "# Create pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "33226c18-863f-4196-9aa9-5fcd36fdab9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-27 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: black;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-27 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-27 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-27 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-27 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-27 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-27 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: block;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-27 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-27 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-27 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-27 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-27 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-27 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-27 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-27 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-27 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-27 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-27 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-27 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 1ex;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-27 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-27 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-27 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-27\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>Pipeline(steps=[(&#x27;Feature selection&#x27;,\n",
       "                 ColumnTransformer(remainder=&#x27;passthrough&#x27;,\n",
       "                                   transformers=[(&#x27;Gene expression&#x27;,\n",
       "                                                  Pipeline(steps=[(&#x27;Non-zero &#x27;\n",
       "                                                                   &#x27;variance&#x27;,\n",
       "                                                                   VarianceSelector()),\n",
       "                                                                  (&#x27;Log1p&#x27;,\n",
       "                                                                   Log1pTransform()),\n",
       "                                                                  (&#x27;Standard &#x27;\n",
       "                                                                   &#x27;scaling&#x27;,\n",
       "                                                                   StandardTransform()),\n",
       "                                                                  (&#x27;Cox &#x27;\n",
       "                                                                   &#x27;ElasticNet&#x27;,\n",
       "                                                                   CoxnetSelector(coef_threshold=0.05))]),\n",
       "                                                  &lt;sklearn.compose._column_transformer.make_column_select...\n",
       "                                                  &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454ca550&gt;),\n",
       "                                                 (&#x27;Gistic copy number&#x27;,\n",
       "                                                  PCATransform(prefix=&#x27;Feature_CN_gistic&#x27;),\n",
       "                                                  &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454ca9a0&gt;),\n",
       "                                                 (&#x27;FISH copy number&#x27;,\n",
       "                                                  PCATransform(prefix=&#x27;Feature_CN_fish&#x27;),\n",
       "                                                  &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454cac40&gt;)]))])</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-586\" type=\"checkbox\" ><label for=\"sk-estimator-id-586\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">&nbsp;&nbsp;Pipeline<a class=\"sk-estimator-doc-link \" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.5/modules/generated/sklearn.pipeline.Pipeline.html\">?<span>Documentation for Pipeline</span></a><span class=\"sk-estimator-doc-link \">i<span>Not fitted</span></span></label><div class=\"sk-toggleable__content \"><pre>Pipeline(steps=[(&#x27;Feature selection&#x27;,\n",
       "                 ColumnTransformer(remainder=&#x27;passthrough&#x27;,\n",
       "                                   transformers=[(&#x27;Gene expression&#x27;,\n",
       "                                                  Pipeline(steps=[(&#x27;Non-zero &#x27;\n",
       "                                                                   &#x27;variance&#x27;,\n",
       "                                                                   VarianceSelector()),\n",
       "                                                                  (&#x27;Log1p&#x27;,\n",
       "                                                                   Log1pTransform()),\n",
       "                                                                  (&#x27;Standard &#x27;\n",
       "                                                                   &#x27;scaling&#x27;,\n",
       "                                                                   StandardTransform()),\n",
       "                                                                  (&#x27;Cox &#x27;\n",
       "                                                                   &#x27;ElasticNet&#x27;,\n",
       "                                                                   CoxnetSelector(coef_threshold=0.05))]),\n",
       "                                                  &lt;sklearn.compose._column_transformer.make_column_select...\n",
       "                                                  &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454ca550&gt;),\n",
       "                                                 (&#x27;Gistic copy number&#x27;,\n",
       "                                                  PCATransform(prefix=&#x27;Feature_CN_gistic&#x27;),\n",
       "                                                  &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454ca9a0&gt;),\n",
       "                                                 (&#x27;FISH copy number&#x27;,\n",
       "                                                  PCATransform(prefix=&#x27;Feature_CN_fish&#x27;),\n",
       "                                                  &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454cac40&gt;)]))])</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-587\" type=\"checkbox\" ><label for=\"sk-estimator-id-587\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">&nbsp;Feature selection: ColumnTransformer<a class=\"sk-estimator-doc-link \" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.5/modules/generated/sklearn.compose.ColumnTransformer.html\">?<span>Documentation for Feature selection: ColumnTransformer</span></a></label><div class=\"sk-toggleable__content \"><pre>ColumnTransformer(remainder=&#x27;passthrough&#x27;,\n",
       "                  transformers=[(&#x27;Gene expression&#x27;,\n",
       "                                 Pipeline(steps=[(&#x27;Non-zero variance&#x27;,\n",
       "                                                  VarianceSelector()),\n",
       "                                                 (&#x27;Log1p&#x27;, Log1pTransform()),\n",
       "                                                 (&#x27;Standard scaling&#x27;,\n",
       "                                                  StandardTransform()),\n",
       "                                                 (&#x27;Cox ElasticNet&#x27;,\n",
       "                                                  CoxnetSelector(coef_threshold=0.05))]),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454c3460&gt;),\n",
       "                                (&#x27;Gene copy numbe...\n",
       "                                                 (&#x27;Log1p&#x27;, Log1pTransform()),\n",
       "                                                 (&#x27;Standard scaling&#x27;,\n",
       "                                                  StandardTransform()),\n",
       "                                                 (&#x27;Cox LASSO&#x27;,\n",
       "                                                  CoxnetSelector(coef_threshold=0.1))]),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454c3220&gt;),\n",
       "                                (&#x27;Clinical&#x27;,\n",
       "                                 Pipeline(steps=[(&#x27;Scale age&#x27;,\n",
       "                                                  StandardTransform(cols=[&#x27;Feature_clin_D_PT_age&#x27;]))]),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454c31c0&gt;)])</pre></div> </div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-588\" type=\"checkbox\" ><label for=\"sk-estimator-id-588\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">Gene expression</label><div class=\"sk-toggleable__content \"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454c3460&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-589\" type=\"checkbox\" ><label for=\"sk-estimator-id-589\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">VarianceSelector</label><div class=\"sk-toggleable__content \"><pre>VarianceSelector()</pre></div> </div></div><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-590\" type=\"checkbox\" ><label for=\"sk-estimator-id-590\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">Log1pTransform</label><div class=\"sk-toggleable__content \"><pre>Log1pTransform()</pre></div> </div></div><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-591\" type=\"checkbox\" ><label for=\"sk-estimator-id-591\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">StandardTransform</label><div class=\"sk-toggleable__content \"><pre>StandardTransform()</pre></div> </div></div><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-592\" type=\"checkbox\" ><label for=\"sk-estimator-id-592\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">CoxnetSelector</label><div class=\"sk-toggleable__content \"><pre>CoxnetSelector(coef_threshold=0.05)</pre></div> </div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-593\" type=\"checkbox\" ><label for=\"sk-estimator-id-593\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">Gene copy number</label><div class=\"sk-toggleable__content \"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454c3430&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-594\" type=\"checkbox\" ><label for=\"sk-estimator-id-594\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">VarianceSelector</label><div class=\"sk-toggleable__content \"><pre>VarianceSelector()</pre></div> </div></div><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-595\" type=\"checkbox\" ><label for=\"sk-estimator-id-595\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">CoxnetSelector</label><div class=\"sk-toggleable__content \"><pre>CoxnetSelector(coef_threshold=0.05)</pre></div> </div></div><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-596\" type=\"checkbox\" ><label for=\"sk-estimator-id-596\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">CorrelationSelector</label><div class=\"sk-toggleable__content \"><pre>CorrelationSelector(threshold=0.9)</pre></div> </div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-597\" type=\"checkbox\" ><label for=\"sk-estimator-id-597\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">Gistic copy number</label><div class=\"sk-toggleable__content \"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454c32e0&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-598\" type=\"checkbox\" ><label for=\"sk-estimator-id-598\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">VarianceSelector</label><div class=\"sk-toggleable__content \"><pre>VarianceSelector()</pre></div> </div></div><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-599\" type=\"checkbox\" ><label for=\"sk-estimator-id-599\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">CoxnetSelector</label><div class=\"sk-toggleable__content \"><pre>CoxnetSelector(coef_threshold=0.1)</pre></div> </div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-600\" type=\"checkbox\" ><label for=\"sk-estimator-id-600\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">FISH copy number</label><div class=\"sk-toggleable__content \"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454c3280&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-601\" type=\"checkbox\" ><label for=\"sk-estimator-id-601\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">VarianceSelector</label><div class=\"sk-toggleable__content \"><pre>VarianceSelector()</pre></div> </div></div><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-602\" type=\"checkbox\" ><label for=\"sk-estimator-id-602\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">CoxnetSelector</label><div class=\"sk-toggleable__content \"><pre>CoxnetSelector(coef_threshold=0.1)</pre></div> </div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-603\" type=\"checkbox\" ><label for=\"sk-estimator-id-603\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">Mutation signatures</label><div class=\"sk-toggleable__content \"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454c3220&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-604\" type=\"checkbox\" ><label for=\"sk-estimator-id-604\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">VarianceSelector</label><div class=\"sk-toggleable__content \"><pre>VarianceSelector()</pre></div> </div></div><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-605\" type=\"checkbox\" ><label for=\"sk-estimator-id-605\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">Log1pTransform</label><div class=\"sk-toggleable__content \"><pre>Log1pTransform()</pre></div> </div></div><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-606\" type=\"checkbox\" ><label for=\"sk-estimator-id-606\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">StandardTransform</label><div class=\"sk-toggleable__content \"><pre>StandardTransform()</pre></div> </div></div><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-607\" type=\"checkbox\" ><label for=\"sk-estimator-id-607\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">CoxnetSelector</label><div class=\"sk-toggleable__content \"><pre>CoxnetSelector(coef_threshold=0.1)</pre></div> </div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-608\" type=\"checkbox\" ><label for=\"sk-estimator-id-608\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">Clinical</label><div class=\"sk-toggleable__content \"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454c31c0&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-609\" type=\"checkbox\" ><label for=\"sk-estimator-id-609\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">StandardTransform</label><div class=\"sk-toggleable__content \"><pre>StandardTransform(cols=[&#x27;Feature_clin_D_PT_age&#x27;])</pre></div> </div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-610\" type=\"checkbox\" ><label for=\"sk-estimator-id-610\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">remainder</label><div class=\"sk-toggleable__content \"><pre></pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-611\" type=\"checkbox\" ><label for=\"sk-estimator-id-611\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">passthrough</label><div class=\"sk-toggleable__content \"><pre>passthrough</pre></div> </div></div></div></div></div></div></div><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-612\" type=\"checkbox\" ><label for=\"sk-estimator-id-612\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">&nbsp;Joint imputation: ColumnTransformer<a class=\"sk-estimator-doc-link \" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.5/modules/generated/sklearn.compose.ColumnTransformer.html\">?<span>Documentation for Joint imputation: ColumnTransformer</span></a></label><div class=\"sk-toggleable__content \"><pre>ColumnTransformer(transformers=[(&#x27;Continuous variables&#x27;,\n",
       "                                 IterativeImputer(estimator=RandomForestRegressor(),\n",
       "                                                  max_iter=100,\n",
       "                                                  n_nearest_features=10,\n",
       "                                                  skip_complete=True,\n",
       "                                                  tol=0.01),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454c30a0&gt;),\n",
       "                                (&#x27;Categorical variables&#x27;,\n",
       "                                 IterativeImputer(estimator=RandomForestClassifier(),\n",
       "                                                  initial_strategy=&#x27;most_frequent&#x27;,\n",
       "                                                  max_iter=100,\n",
       "                                                  n_nearest_features=10,\n",
       "                                                  skip_complete=True,\n",
       "                                                  tol=0.01),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454ca760&gt;)])</pre></div> </div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-613\" type=\"checkbox\" ><label for=\"sk-estimator-id-613\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">Continuous variables</label><div class=\"sk-toggleable__content \"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454c30a0&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-614\" type=\"checkbox\" ><label for=\"sk-estimator-id-614\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">estimator: RandomForestRegressor</label><div class=\"sk-toggleable__content \"><pre>RandomForestRegressor()</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-615\" type=\"checkbox\" ><label for=\"sk-estimator-id-615\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">&nbsp;RandomForestRegressor<a class=\"sk-estimator-doc-link \" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.5/modules/generated/sklearn.ensemble.RandomForestRegressor.html\">?<span>Documentation for RandomForestRegressor</span></a></label><div class=\"sk-toggleable__content \"><pre>RandomForestRegressor()</pre></div> </div></div></div></div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-616\" type=\"checkbox\" ><label for=\"sk-estimator-id-616\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">Categorical variables</label><div class=\"sk-toggleable__content \"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454ca760&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-617\" type=\"checkbox\" ><label for=\"sk-estimator-id-617\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">estimator: RandomForestClassifier</label><div class=\"sk-toggleable__content \"><pre>RandomForestClassifier()</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-618\" type=\"checkbox\" ><label for=\"sk-estimator-id-618\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">&nbsp;RandomForestClassifier<a class=\"sk-estimator-doc-link \" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.5/modules/generated/sklearn.ensemble.RandomForestClassifier.html\">?<span>Documentation for RandomForestClassifier</span></a></label><div class=\"sk-toggleable__content \"><pre>RandomForestClassifier()</pre></div> </div></div></div></div></div></div></div></div></div></div></div></div><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-619\" type=\"checkbox\" ><label for=\"sk-estimator-id-619\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">&nbsp;PCA: ColumnTransformer<a class=\"sk-estimator-doc-link \" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.5/modules/generated/sklearn.compose.ColumnTransformer.html\">?<span>Documentation for PCA: ColumnTransformer</span></a></label><div class=\"sk-toggleable__content \"><pre>ColumnTransformer(remainder=&#x27;passthrough&#x27;,\n",
       "                  transformers=[(&#x27;Gene expression&#x27;,\n",
       "                                 PCATransform(prefix=&#x27;Feature_exp&#x27;),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454cab20&gt;),\n",
       "                                (&#x27;Gene copy number&#x27;,\n",
       "                                 PCATransform(prefix=&#x27;Feature_CN_gene&#x27;),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454ca550&gt;),\n",
       "                                (&#x27;Gistic copy number&#x27;,\n",
       "                                 PCATransform(prefix=&#x27;Feature_CN_gistic&#x27;),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454ca9a0&gt;),\n",
       "                                (&#x27;FISH copy number&#x27;,\n",
       "                                 PCATransform(prefix=&#x27;Feature_CN_fish&#x27;),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454cac40&gt;)])</pre></div> </div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-620\" type=\"checkbox\" ><label for=\"sk-estimator-id-620\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">Gene expression</label><div class=\"sk-toggleable__content \"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454cab20&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-621\" type=\"checkbox\" ><label for=\"sk-estimator-id-621\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">PCATransform</label><div class=\"sk-toggleable__content \"><pre>PCATransform(prefix=&#x27;Feature_exp&#x27;)</pre></div> </div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-622\" type=\"checkbox\" ><label for=\"sk-estimator-id-622\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">Gene copy number</label><div class=\"sk-toggleable__content \"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454ca550&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-623\" type=\"checkbox\" ><label for=\"sk-estimator-id-623\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">PCATransform</label><div class=\"sk-toggleable__content \"><pre>PCATransform(prefix=&#x27;Feature_CN_gene&#x27;)</pre></div> </div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-624\" type=\"checkbox\" ><label for=\"sk-estimator-id-624\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">Gistic copy number</label><div class=\"sk-toggleable__content \"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454ca9a0&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-625\" type=\"checkbox\" ><label for=\"sk-estimator-id-625\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">PCATransform</label><div class=\"sk-toggleable__content \"><pre>PCATransform(prefix=&#x27;Feature_CN_gistic&#x27;)</pre></div> </div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-626\" type=\"checkbox\" ><label for=\"sk-estimator-id-626\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">FISH copy number</label><div class=\"sk-toggleable__content \"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x154c454cac40&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-627\" type=\"checkbox\" ><label for=\"sk-estimator-id-627\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">PCATransform</label><div class=\"sk-toggleable__content \"><pre>PCATransform(prefix=&#x27;Feature_CN_fish&#x27;)</pre></div> </div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-628\" type=\"checkbox\" ><label for=\"sk-estimator-id-628\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">remainder</label><div class=\"sk-toggleable__content \"><pre></pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator  sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-629\" type=\"checkbox\" ><label for=\"sk-estimator-id-629\" class=\"sk-toggleable__label  sk-toggleable__label-arrow \">passthrough</label><div class=\"sk-toggleable__content \"><pre>passthrough</pre></div> </div></div></div></div></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "Pipeline(steps=[('Feature selection',\n",
       "                 ColumnTransformer(remainder='passthrough',\n",
       "                                   transformers=[('Gene expression',\n",
       "                                                  Pipeline(steps=[('Non-zero '\n",
       "                                                                   'variance',\n",
       "                                                                   VarianceSelector()),\n",
       "                                                                  ('Log1p',\n",
       "                                                                   Log1pTransform()),\n",
       "                                                                  ('Standard '\n",
       "                                                                   'scaling',\n",
       "                                                                   StandardTransform()),\n",
       "                                                                  ('Cox '\n",
       "                                                                   'ElasticNet',\n",
       "                                                                   CoxnetSelector(coef_threshold=0.05))]),\n",
       "                                                  <sklearn.compose._column_transformer.make_column_select...\n",
       "                                                  <sklearn.compose._column_transformer.make_column_selector object at 0x154c454ca550>),\n",
       "                                                 ('Gistic copy number',\n",
       "                                                  PCATransform(prefix='Feature_CN_gistic'),\n",
       "                                                  <sklearn.compose._column_transformer.make_column_selector object at 0x154c454ca9a0>),\n",
       "                                                 ('FISH copy number',\n",
       "                                                  PCATransform(prefix='Feature_CN_fish'),\n",
       "                                                  <sklearn.compose._column_transformer.make_column_selector object at 0x154c454cac40>)]))])"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transformer_gene_exp = Pipeline([\n",
    "    ('Non-zero variance', VarianceSelector(threshold=0)),\n",
    "    ('Log1p', Log1pTransform()),\n",
    "    ('Standard scaling', StandardTransform()),\n",
    "    ('Cox ElasticNet', CoxnetSelector(l1_ratio=0.5, coef_threshold=0.05)),\n",
    "])\n",
    "\n",
    "transformer_gene_cn = Pipeline([\n",
    "    ('Non-zero variance', VarianceSelector(threshold=0)),\n",
    "    ('Coxnet', CoxnetSelector(l1_ratio=0.5, coef_threshold=0.05)),\n",
    "    ('Uncorrelated', CorrelationSelector(threshold=0.9)),\n",
    "])\n",
    "\n",
    "transformer_gistic = Pipeline([\n",
    "    ('Non-zero variance', VarianceSelector(threshold=0)),\n",
    "    ('Coxnet', CoxnetSelector(l1_ratio=0.5, coef_threshold = 0.1)),\n",
    "])\n",
    "\n",
    "transformer_sbs = Pipeline([\n",
    "    ('Non-zero variance', VarianceSelector(threshold=1)),\n",
    "    ('Log1p', Log1pTransform()),\n",
    "    ('Standard scaling', StandardTransform()),\n",
    "    ('Cox LASSO', CoxnetSelector(l1_ratio=1, coef_threshold=0.1)),\n",
    "])\n",
    "\n",
    "transformer_fish = Pipeline([\n",
    "    ('Non-zero variance', VarianceSelector(threshold=0)),\n",
    "    ('Coxnet', CoxnetSelector(l1_ratio=0.5, coef_threshold = 0.1)),\n",
    "])\n",
    "\n",
    "transformer_ig = Pipeline([\n",
    "    ('Frequency', FrequencySelector(minfreq=0.01))\n",
    "])\n",
    "\n",
    "transformer_clin = Pipeline([\n",
    "    ('Scale age', StandardTransform(cols=['Feature_clin_D_PT_age']))\n",
    "])\n",
    "\n",
    "transformer = ColumnTransformer([\n",
    "    ('Gene expression', transformer_gene_exp, make_column_selector(pattern='Feature_exp_')),\n",
    "    ('Gene copy number', transformer_gene_cn, make_column_selector(pattern='Feature_CNA_ENSG')),\n",
    "    ('Gistic copy number', transformer_gistic, make_column_selector(pattern='Feature_CNA_(Amp|Del)')),\n",
    "    ('FISH copy number', transformer_fish, make_column_selector(pattern='Feature_fish')),\n",
    "    ('Mutation signatures', transformer_sbs, make_column_selector(pattern='Feature_SBS')),\n",
    "    ('Clinical', transformer_clin, make_column_selector(pattern='Feature_clin')),\n",
    "], remainder='passthrough').set_output(transform=\"pandas\")\n",
    "\n",
    "imputer_params = {\n",
    "    'n_nearest_features': 10, \n",
    "    'max_iter': 100,\n",
    "    'tol':1e-2,\n",
    "    'skip_complete':True\n",
    "}\n",
    "\n",
    "RegressionImputer = IterativeImputer(estimator=RandomForestRegressor(), initial_strategy='mean', **imputer_params)\n",
    "ClassificationImputer = IterativeImputer(estimator=RandomForestClassifier(), initial_strategy='most_frequent', **imputer_params)\n",
    "\n",
    "imputer = ColumnTransformer([\n",
    "    ('Continuous variables', RegressionImputer, make_column_selector(pattern='(Feature_(exp|clin_D_PT_age|SBS)|pca)')),\n",
    "    ('Categorical variables', ClassificationImputer, make_column_selector(pattern='Feature_(?!exp|clin_D_PT_age|SBS)'))\n",
    "], remainder='drop').set_output(transform=\"pandas\")\n",
    "\n",
    "pca = ColumnTransformer([\n",
    "    ('Gene expression', PCATransform(prefix='Feature_exp',n_components=10), make_column_selector(pattern='Feature_exp')),\n",
    "    ('Gene copy number', PCATransform(prefix='Feature_CN_gene',n_components=10), make_column_selector(pattern='Feature_CNA_ENSG')),\n",
    "    ('Gistic copy number', PCATransform(prefix='Feature_CN_gistic',n_components=10), make_column_selector(pattern='Feature_CNA_(RNASeq|SeqWGS)')),\n",
    "    ('FISH copy number', PCATransform(prefix='Feature_CN_fish',n_components=10), make_column_selector(pattern='Feature_fish)')),\n",
    "], remainder='passthrough').set_output(transform=\"pandas\")\n",
    "\n",
    "pipeline = Pipeline([\n",
    "    ('Feature selection', transformer),\n",
    "    ('Joint imputation', imputer),\n",
    "    ('PCA', pca)\n",
    "])\n",
    "pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "68c26d32",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/users/nus/e1083772/.localpython/lib/python3.9/site-packages/sklearn/compose/_column_transformer.py:1553: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  cols = cols[cols.str.contains(self.pattern, regex=True)]\n",
      "/home/users/nus/e1083772/.localpython/lib/python3.9/site-packages/sklearn/compose/_column_transformer.py:1553: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  cols = cols[cols.str.contains(self.pattern, regex=True)]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[134], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m out \u001b[38;5;241m=\u001b[39m \u001b[43mpipeline\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit_transform\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfeatures\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/base.py:1473\u001b[0m, in \u001b[0;36m_fit_context.<locals>.decorator.<locals>.wrapper\u001b[0;34m(estimator, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1466\u001b[0m     estimator\u001b[38;5;241m.\u001b[39m_validate_params()\n\u001b[1;32m   1468\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m config_context(\n\u001b[1;32m   1469\u001b[0m     skip_parameter_validation\u001b[38;5;241m=\u001b[39m(\n\u001b[1;32m   1470\u001b[0m         prefer_skip_nested_validation \u001b[38;5;129;01mor\u001b[39;00m global_skip_validation\n\u001b[1;32m   1471\u001b[0m     )\n\u001b[1;32m   1472\u001b[0m ):\n\u001b[0;32m-> 1473\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfit_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mestimator\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/pipeline.py:541\u001b[0m, in \u001b[0;36mPipeline.fit_transform\u001b[0;34m(self, X, y, **params)\u001b[0m\n\u001b[1;32m    539\u001b[0m last_step_params \u001b[38;5;241m=\u001b[39m routed_params[\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msteps[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m][\u001b[38;5;241m0\u001b[39m]]\n\u001b[1;32m    540\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(last_step, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfit_transform\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[0;32m--> 541\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mlast_step\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit_transform\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    542\u001b[0m \u001b[43m        \u001b[49m\u001b[43mXt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mlast_step_params\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfit_transform\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m    543\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    544\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    545\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m last_step\u001b[38;5;241m.\u001b[39mfit(Xt, y, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mlast_step_params[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfit\u001b[39m\u001b[38;5;124m\"\u001b[39m])\u001b[38;5;241m.\u001b[39mtransform(\n\u001b[1;32m    546\u001b[0m         Xt, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mlast_step_params[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtransform\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m    547\u001b[0m     )\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/utils/_set_output.py:316\u001b[0m, in \u001b[0;36m_wrap_method_output.<locals>.wrapped\u001b[0;34m(self, X, *args, **kwargs)\u001b[0m\n\u001b[1;32m    314\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(f)\n\u001b[1;32m    315\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrapped\u001b[39m(\u001b[38;5;28mself\u001b[39m, X, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m--> 316\u001b[0m     data_to_wrap \u001b[38;5;241m=\u001b[39m \u001b[43mf\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    317\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data_to_wrap, \u001b[38;5;28mtuple\u001b[39m):\n\u001b[1;32m    318\u001b[0m         \u001b[38;5;66;03m# only wrap the first output for cross decomposition\u001b[39;00m\n\u001b[1;32m    319\u001b[0m         return_tuple \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    320\u001b[0m             _wrap_data_with_container(method, data_to_wrap[\u001b[38;5;241m0\u001b[39m], X, \u001b[38;5;28mself\u001b[39m),\n\u001b[1;32m    321\u001b[0m             \u001b[38;5;241m*\u001b[39mdata_to_wrap[\u001b[38;5;241m1\u001b[39m:],\n\u001b[1;32m    322\u001b[0m         )\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/base.py:1473\u001b[0m, in \u001b[0;36m_fit_context.<locals>.decorator.<locals>.wrapper\u001b[0;34m(estimator, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1466\u001b[0m     estimator\u001b[38;5;241m.\u001b[39m_validate_params()\n\u001b[1;32m   1468\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m config_context(\n\u001b[1;32m   1469\u001b[0m     skip_parameter_validation\u001b[38;5;241m=\u001b[39m(\n\u001b[1;32m   1470\u001b[0m         prefer_skip_nested_validation \u001b[38;5;129;01mor\u001b[39;00m global_skip_validation\n\u001b[1;32m   1471\u001b[0m     )\n\u001b[1;32m   1472\u001b[0m ):\n\u001b[0;32m-> 1473\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfit_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mestimator\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/compose/_column_transformer.py:976\u001b[0m, in \u001b[0;36mColumnTransformer.fit_transform\u001b[0;34m(self, X, y, **params)\u001b[0m\n\u001b[1;32m    973\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    974\u001b[0m     routed_params \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_empty_routing()\n\u001b[0;32m--> 976\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_func_on_transformers\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    977\u001b[0m \u001b[43m    \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    978\u001b[0m \u001b[43m    \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    979\u001b[0m \u001b[43m    \u001b[49m\u001b[43m_fit_transform_one\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    980\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcolumn_as_labels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    981\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrouted_params\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrouted_params\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    982\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    984\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m result:\n\u001b[1;32m    985\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_update_fitted_transformers([])\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/compose/_column_transformer.py:885\u001b[0m, in \u001b[0;36mColumnTransformer._call_func_on_transformers\u001b[0;34m(self, X, y, func, column_as_labels, routed_params)\u001b[0m\n\u001b[1;32m    873\u001b[0m             extra_args \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m    874\u001b[0m         jobs\u001b[38;5;241m.\u001b[39mappend(\n\u001b[1;32m    875\u001b[0m             delayed(func)(\n\u001b[1;32m    876\u001b[0m                 transformer\u001b[38;5;241m=\u001b[39mclone(trans) \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m fitted \u001b[38;5;28;01melse\u001b[39;00m trans,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    882\u001b[0m             )\n\u001b[1;32m    883\u001b[0m         )\n\u001b[0;32m--> 885\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mParallel\u001b[49m\u001b[43m(\u001b[49m\u001b[43mn_jobs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mn_jobs\u001b[49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\u001b[43mjobs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    887\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    888\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mExpected 2D array, got 1D array instead\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mstr\u001b[39m(e):\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/utils/parallel.py:74\u001b[0m, in \u001b[0;36mParallel.__call__\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m     69\u001b[0m config \u001b[38;5;241m=\u001b[39m get_config()\n\u001b[1;32m     70\u001b[0m iterable_with_config \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m     71\u001b[0m     (_with_config(delayed_func, config), args, kwargs)\n\u001b[1;32m     72\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m delayed_func, args, kwargs \u001b[38;5;129;01min\u001b[39;00m iterable\n\u001b[1;32m     73\u001b[0m )\n\u001b[0;32m---> 74\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__call__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43miterable_with_config\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/joblib/parallel.py:1918\u001b[0m, in \u001b[0;36mParallel.__call__\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m   1916\u001b[0m     output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_sequential_output(iterable)\n\u001b[1;32m   1917\u001b[0m     \u001b[38;5;28mnext\u001b[39m(output)\n\u001b[0;32m-> 1918\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m output \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturn_generator \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;43mlist\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43moutput\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1920\u001b[0m \u001b[38;5;66;03m# Let's create an ID that uniquely identifies the current call. If the\u001b[39;00m\n\u001b[1;32m   1921\u001b[0m \u001b[38;5;66;03m# call is interrupted early and that the same instance is immediately\u001b[39;00m\n\u001b[1;32m   1922\u001b[0m \u001b[38;5;66;03m# re-used, this id will be used to prevent workers that were\u001b[39;00m\n\u001b[1;32m   1923\u001b[0m \u001b[38;5;66;03m# concurrently finalizing a task from the previous call to run the\u001b[39;00m\n\u001b[1;32m   1924\u001b[0m \u001b[38;5;66;03m# callback.\u001b[39;00m\n\u001b[1;32m   1925\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_lock:\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/joblib/parallel.py:1847\u001b[0m, in \u001b[0;36mParallel._get_sequential_output\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m   1845\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_dispatched_batches \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m   1846\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_dispatched_tasks \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[0;32m-> 1847\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1848\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_completed_tasks \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m   1849\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprint_progress()\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/utils/parallel.py:136\u001b[0m, in \u001b[0;36m_FuncWrapper.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    134\u001b[0m     config \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m    135\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m config_context(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mconfig):\n\u001b[0;32m--> 136\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/pipeline.py:1310\u001b[0m, in \u001b[0;36m_fit_transform_one\u001b[0;34m(transformer, X, y, weight, message_clsname, message, params)\u001b[0m\n\u001b[1;32m   1308\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m _print_elapsed_time(message_clsname, message):\n\u001b[1;32m   1309\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(transformer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfit_transform\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[0;32m-> 1310\u001b[0m         res \u001b[38;5;241m=\u001b[39m \u001b[43mtransformer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit_transform\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfit_transform\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m{\u001b[49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1311\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1312\u001b[0m         res \u001b[38;5;241m=\u001b[39m transformer\u001b[38;5;241m.\u001b[39mfit(X, y, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mparams\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfit\u001b[39m\u001b[38;5;124m\"\u001b[39m, {}))\u001b[38;5;241m.\u001b[39mtransform(\n\u001b[1;32m   1313\u001b[0m             X, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mparams\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtransform\u001b[39m\u001b[38;5;124m\"\u001b[39m, {})\n\u001b[1;32m   1314\u001b[0m         )\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/utils/_set_output.py:316\u001b[0m, in \u001b[0;36m_wrap_method_output.<locals>.wrapped\u001b[0;34m(self, X, *args, **kwargs)\u001b[0m\n\u001b[1;32m    314\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(f)\n\u001b[1;32m    315\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrapped\u001b[39m(\u001b[38;5;28mself\u001b[39m, X, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m--> 316\u001b[0m     data_to_wrap \u001b[38;5;241m=\u001b[39m \u001b[43mf\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    317\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data_to_wrap, \u001b[38;5;28mtuple\u001b[39m):\n\u001b[1;32m    318\u001b[0m         \u001b[38;5;66;03m# only wrap the first output for cross decomposition\u001b[39;00m\n\u001b[1;32m    319\u001b[0m         return_tuple \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    320\u001b[0m             _wrap_data_with_container(method, data_to_wrap[\u001b[38;5;241m0\u001b[39m], X, \u001b[38;5;28mself\u001b[39m),\n\u001b[1;32m    321\u001b[0m             \u001b[38;5;241m*\u001b[39mdata_to_wrap[\u001b[38;5;241m1\u001b[39m:],\n\u001b[1;32m    322\u001b[0m         )\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/base.py:1473\u001b[0m, in \u001b[0;36m_fit_context.<locals>.decorator.<locals>.wrapper\u001b[0;34m(estimator, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1466\u001b[0m     estimator\u001b[38;5;241m.\u001b[39m_validate_params()\n\u001b[1;32m   1468\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m config_context(\n\u001b[1;32m   1469\u001b[0m     skip_parameter_validation\u001b[38;5;241m=\u001b[39m(\n\u001b[1;32m   1470\u001b[0m         prefer_skip_nested_validation \u001b[38;5;129;01mor\u001b[39;00m global_skip_validation\n\u001b[1;32m   1471\u001b[0m     )\n\u001b[1;32m   1472\u001b[0m ):\n\u001b[0;32m-> 1473\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfit_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mestimator\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/impute/_iterative.py:789\u001b[0m, in \u001b[0;36mIterativeImputer.fit_transform\u001b[0;34m(self, X, y, **params)\u001b[0m\n\u001b[1;32m    785\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m feat_idx \u001b[38;5;129;01min\u001b[39;00m ordered_idx:\n\u001b[1;32m    786\u001b[0m     neighbor_feat_idx \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_neighbor_feat_idx(\n\u001b[1;32m    787\u001b[0m         n_features, feat_idx, abs_corr_mat\n\u001b[1;32m    788\u001b[0m     )\n\u001b[0;32m--> 789\u001b[0m     Xt, estimator \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_impute_one_feature\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    790\u001b[0m \u001b[43m        \u001b[49m\u001b[43mXt\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    791\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmask_missing_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    792\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfeat_idx\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    793\u001b[0m \u001b[43m        \u001b[49m\u001b[43mneighbor_feat_idx\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    794\u001b[0m \u001b[43m        \u001b[49m\u001b[43mestimator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    795\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfit_mode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    796\u001b[0m \u001b[43m        \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrouted_params\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mestimator\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    797\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    798\u001b[0m     estimator_triplet \u001b[38;5;241m=\u001b[39m _ImputerTriplet(\n\u001b[1;32m    799\u001b[0m         feat_idx, neighbor_feat_idx, estimator\n\u001b[1;32m    800\u001b[0m     )\n\u001b[1;32m    801\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mimputation_sequence_\u001b[38;5;241m.\u001b[39mappend(estimator_triplet)\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/impute/_iterative.py:418\u001b[0m, in \u001b[0;36mIterativeImputer._impute_one_feature\u001b[0;34m(self, X_filled, mask_missing_values, feat_idx, neighbor_feat_idx, estimator, fit_mode, params)\u001b[0m\n\u001b[1;32m    408\u001b[0m     X_train \u001b[38;5;241m=\u001b[39m _safe_indexing(\n\u001b[1;32m    409\u001b[0m         _safe_indexing(X_filled, neighbor_feat_idx, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m),\n\u001b[1;32m    410\u001b[0m         \u001b[38;5;241m~\u001b[39mmissing_row_mask,\n\u001b[1;32m    411\u001b[0m         axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m,\n\u001b[1;32m    412\u001b[0m     )\n\u001b[1;32m    413\u001b[0m     y_train \u001b[38;5;241m=\u001b[39m _safe_indexing(\n\u001b[1;32m    414\u001b[0m         _safe_indexing(X_filled, feat_idx, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m),\n\u001b[1;32m    415\u001b[0m         \u001b[38;5;241m~\u001b[39mmissing_row_mask,\n\u001b[1;32m    416\u001b[0m         axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m,\n\u001b[1;32m    417\u001b[0m     )\n\u001b[0;32m--> 418\u001b[0m     \u001b[43mestimator\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX_train\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my_train\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    420\u001b[0m \u001b[38;5;66;03m# if no missing values, don't predict\u001b[39;00m\n\u001b[1;32m    421\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m np\u001b[38;5;241m.\u001b[39msum(missing_row_mask) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/base.py:1473\u001b[0m, in \u001b[0;36m_fit_context.<locals>.decorator.<locals>.wrapper\u001b[0;34m(estimator, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1466\u001b[0m     estimator\u001b[38;5;241m.\u001b[39m_validate_params()\n\u001b[1;32m   1468\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m config_context(\n\u001b[1;32m   1469\u001b[0m     skip_parameter_validation\u001b[38;5;241m=\u001b[39m(\n\u001b[1;32m   1470\u001b[0m         prefer_skip_nested_validation \u001b[38;5;129;01mor\u001b[39;00m global_skip_validation\n\u001b[1;32m   1471\u001b[0m     )\n\u001b[1;32m   1472\u001b[0m ):\n\u001b[0;32m-> 1473\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfit_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mestimator\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/ensemble/_forest.py:489\u001b[0m, in \u001b[0;36mBaseForest.fit\u001b[0;34m(self, X, y, sample_weight)\u001b[0m\n\u001b[1;32m    478\u001b[0m trees \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m    479\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_estimator(append\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, random_state\u001b[38;5;241m=\u001b[39mrandom_state)\n\u001b[1;32m    480\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(n_more_estimators)\n\u001b[1;32m    481\u001b[0m ]\n\u001b[1;32m    483\u001b[0m \u001b[38;5;66;03m# Parallel loop: we prefer the threading backend as the Cython code\u001b[39;00m\n\u001b[1;32m    484\u001b[0m \u001b[38;5;66;03m# for fitting the trees is internally releasing the Python GIL\u001b[39;00m\n\u001b[1;32m    485\u001b[0m \u001b[38;5;66;03m# making threading more efficient than multiprocessing in\u001b[39;00m\n\u001b[1;32m    486\u001b[0m \u001b[38;5;66;03m# that case. However, for joblib 0.12+ we respect any\u001b[39;00m\n\u001b[1;32m    487\u001b[0m \u001b[38;5;66;03m# parallel_backend contexts set at a higher level,\u001b[39;00m\n\u001b[1;32m    488\u001b[0m \u001b[38;5;66;03m# since correctness does not rely on using threads.\u001b[39;00m\n\u001b[0;32m--> 489\u001b[0m trees \u001b[38;5;241m=\u001b[39m \u001b[43mParallel\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    490\u001b[0m \u001b[43m    \u001b[49m\u001b[43mn_jobs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mn_jobs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    491\u001b[0m \u001b[43m    \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    492\u001b[0m \u001b[43m    \u001b[49m\u001b[43mprefer\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mthreads\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    493\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdelayed\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_parallel_build_trees\u001b[49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    495\u001b[0m \u001b[43m        \u001b[49m\u001b[43mt\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    496\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbootstrap\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    497\u001b[0m \u001b[43m        \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    498\u001b[0m \u001b[43m        \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    499\u001b[0m \u001b[43m        \u001b[49m\u001b[43msample_weight\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    500\u001b[0m \u001b[43m        \u001b[49m\u001b[43mi\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    501\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mtrees\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    502\u001b[0m \u001b[43m        \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    503\u001b[0m \u001b[43m        \u001b[49m\u001b[43mclass_weight\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclass_weight\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    504\u001b[0m \u001b[43m        \u001b[49m\u001b[43mn_samples_bootstrap\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mn_samples_bootstrap\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    505\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmissing_values_in_feature_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmissing_values_in_feature_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    506\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    507\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mi\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mt\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43menumerate\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mtrees\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    508\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    510\u001b[0m \u001b[38;5;66;03m# Collect newly grown trees\u001b[39;00m\n\u001b[1;32m    511\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mestimators_\u001b[38;5;241m.\u001b[39mextend(trees)\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/utils/parallel.py:74\u001b[0m, in \u001b[0;36mParallel.__call__\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m     69\u001b[0m config \u001b[38;5;241m=\u001b[39m get_config()\n\u001b[1;32m     70\u001b[0m iterable_with_config \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m     71\u001b[0m     (_with_config(delayed_func, config), args, kwargs)\n\u001b[1;32m     72\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m delayed_func, args, kwargs \u001b[38;5;129;01min\u001b[39;00m iterable\n\u001b[1;32m     73\u001b[0m )\n\u001b[0;32m---> 74\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__call__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43miterable_with_config\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/joblib/parallel.py:1918\u001b[0m, in \u001b[0;36mParallel.__call__\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m   1916\u001b[0m     output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_sequential_output(iterable)\n\u001b[1;32m   1917\u001b[0m     \u001b[38;5;28mnext\u001b[39m(output)\n\u001b[0;32m-> 1918\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m output \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturn_generator \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;43mlist\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43moutput\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1920\u001b[0m \u001b[38;5;66;03m# Let's create an ID that uniquely identifies the current call. If the\u001b[39;00m\n\u001b[1;32m   1921\u001b[0m \u001b[38;5;66;03m# call is interrupted early and that the same instance is immediately\u001b[39;00m\n\u001b[1;32m   1922\u001b[0m \u001b[38;5;66;03m# re-used, this id will be used to prevent workers that were\u001b[39;00m\n\u001b[1;32m   1923\u001b[0m \u001b[38;5;66;03m# concurrently finalizing a task from the previous call to run the\u001b[39;00m\n\u001b[1;32m   1924\u001b[0m \u001b[38;5;66;03m# callback.\u001b[39;00m\n\u001b[1;32m   1925\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_lock:\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/joblib/parallel.py:1847\u001b[0m, in \u001b[0;36mParallel._get_sequential_output\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m   1845\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_dispatched_batches \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m   1846\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_dispatched_tasks \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[0;32m-> 1847\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1848\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_completed_tasks \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m   1849\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprint_progress()\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/utils/parallel.py:136\u001b[0m, in \u001b[0;36m_FuncWrapper.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    134\u001b[0m     config \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m    135\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m config_context(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mconfig):\n\u001b[0;32m--> 136\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/ensemble/_forest.py:179\u001b[0m, in \u001b[0;36m_parallel_build_trees\u001b[0;34m(tree, bootstrap, X, y, sample_weight, tree_idx, n_trees, verbose, class_weight, n_samples_bootstrap, missing_values_in_feature_mask)\u001b[0m\n\u001b[1;32m    176\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    177\u001b[0m     curr_sample_weight \u001b[38;5;241m=\u001b[39m sample_weight\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[0;32m--> 179\u001b[0m indices \u001b[38;5;241m=\u001b[39m \u001b[43m_generate_sample_indices\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    180\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtree\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrandom_state\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_samples\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_samples_bootstrap\u001b[49m\n\u001b[1;32m    181\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    182\u001b[0m sample_counts \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mbincount(indices, minlength\u001b[38;5;241m=\u001b[39mn_samples)\n\u001b[1;32m    183\u001b[0m curr_sample_weight \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m=\u001b[39m sample_counts\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/ensemble/_forest.py:132\u001b[0m, in \u001b[0;36m_generate_sample_indices\u001b[0;34m(random_state, n_samples, n_samples_bootstrap)\u001b[0m\n\u001b[1;32m    128\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_generate_sample_indices\u001b[39m(random_state, n_samples, n_samples_bootstrap):\n\u001b[1;32m    129\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    130\u001b[0m \u001b[38;5;124;03m    Private function used to _parallel_build_trees function.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 132\u001b[0m     random_instance \u001b[38;5;241m=\u001b[39m \u001b[43mcheck_random_state\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrandom_state\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    133\u001b[0m     sample_indices \u001b[38;5;241m=\u001b[39m random_instance\u001b[38;5;241m.\u001b[39mrandint(\n\u001b[1;32m    134\u001b[0m         \u001b[38;5;241m0\u001b[39m, n_samples, n_samples_bootstrap, dtype\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39mint32\n\u001b[1;32m    135\u001b[0m     )\n\u001b[1;32m    137\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m sample_indices\n",
      "File \u001b[0;32m~/.localpython/lib/python3.9/site-packages/sklearn/utils/validation.py:1436\u001b[0m, in \u001b[0;36mcheck_random_state\u001b[0;34m(seed)\u001b[0m\n\u001b[1;32m   1434\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m np\u001b[38;5;241m.\u001b[39mrandom\u001b[38;5;241m.\u001b[39mmtrand\u001b[38;5;241m.\u001b[39m_rand\n\u001b[1;32m   1435\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(seed, numbers\u001b[38;5;241m.\u001b[39mIntegral):\n\u001b[0;32m-> 1436\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrandom\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mRandomState\u001b[49m\u001b[43m(\u001b[49m\u001b[43mseed\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1437\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(seed, np\u001b[38;5;241m.\u001b[39mrandom\u001b[38;5;241m.\u001b[39mRandomState):\n\u001b[1;32m   1438\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m seed\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "out = pipeline.fit_transform(features, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b14538f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene expression__Feature_exp_pca0</th>\n",
       "      <th>Gene expression__Feature_exp_pca1</th>\n",
       "      <th>Gene expression__Feature_exp_pca2</th>\n",
       "      <th>Gene expression__Feature_exp_pca3</th>\n",
       "      <th>Gene expression__Feature_exp_pca4</th>\n",
       "      <th>Gene expression__Feature_exp_pca5</th>\n",
       "      <th>Gene expression__Feature_exp_pca6</th>\n",
       "      <th>Gene expression__Feature_exp_pca7</th>\n",
       "      <th>Gene expression__Feature_exp_pca8</th>\n",
       "      <th>Gene expression__Feature_exp_pca9</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene expression__Feature_exp_pca655</th>\n",
       "      <th>Gene expression__Feature_exp_pca656</th>\n",
       "      <th>Gene expression__Feature_exp_pca657</th>\n",
       "      <th>Gene expression__Feature_exp_pca658</th>\n",
       "      <th>Gene expression__Feature_exp_pca659</th>\n",
       "      <th>Gene expression__Feature_exp_pca660</th>\n",
       "      <th>Gene expression__Feature_exp_pca661</th>\n",
       "      <th>Gene expression__Feature_exp_pca662</th>\n",
       "      <th>Gene expression__Feature_exp_pca663</th>\n",
       "      <th>Gene expression__Feature_exp_pca664</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PUBLIC_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MMRF_1007</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_1011</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_1013</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_1014</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_1016</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_2846</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_2847</th>\n",
       "      <td>0.115155</td>\n",
       "      <td>-0.262712</td>\n",
       "      <td>9.547609</td>\n",
       "      <td>0.174491</td>\n",
       "      <td>-1.700913</td>\n",
       "      <td>-2.594641</td>\n",
       "      <td>0.453928</td>\n",
       "      <td>2.332060</td>\n",
       "      <td>3.599387</td>\n",
       "      <td>3.843321</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001026</td>\n",
       "      <td>0.007154</td>\n",
       "      <td>-0.035740</td>\n",
       "      <td>-0.027436</td>\n",
       "      <td>0.005234</td>\n",
       "      <td>0.037007</td>\n",
       "      <td>-0.010941</td>\n",
       "      <td>0.129442</td>\n",
       "      <td>-0.037312</td>\n",
       "      <td>1.899866e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_2848</th>\n",
       "      <td>-4.057280</td>\n",
       "      <td>5.160718</td>\n",
       "      <td>7.087989</td>\n",
       "      <td>-0.509473</td>\n",
       "      <td>0.181885</td>\n",
       "      <td>2.476096</td>\n",
       "      <td>-2.930544</td>\n",
       "      <td>2.171186</td>\n",
       "      <td>-0.907540</td>\n",
       "      <td>1.734954</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.010497</td>\n",
       "      <td>-0.012944</td>\n",
       "      <td>0.043653</td>\n",
       "      <td>0.027963</td>\n",
       "      <td>0.014943</td>\n",
       "      <td>0.000232</td>\n",
       "      <td>-0.031268</td>\n",
       "      <td>0.041034</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>7.341321e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_2851</th>\n",
       "      <td>-3.540592</td>\n",
       "      <td>-1.420694</td>\n",
       "      <td>6.074475</td>\n",
       "      <td>1.946850</td>\n",
       "      <td>2.362985</td>\n",
       "      <td>-1.068620</td>\n",
       "      <td>-0.522922</td>\n",
       "      <td>-0.722730</td>\n",
       "      <td>-1.051358</td>\n",
       "      <td>-1.105945</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.010703</td>\n",
       "      <td>-0.005334</td>\n",
       "      <td>0.057544</td>\n",
       "      <td>0.053831</td>\n",
       "      <td>0.067620</td>\n",
       "      <td>0.005861</td>\n",
       "      <td>-0.074586</td>\n",
       "      <td>0.006660</td>\n",
       "      <td>-0.009274</td>\n",
       "      <td>-6.120133e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_2853</th>\n",
       "      <td>1.753795</td>\n",
       "      <td>-0.998007</td>\n",
       "      <td>3.684521</td>\n",
       "      <td>-4.461104</td>\n",
       "      <td>-7.864048</td>\n",
       "      <td>0.810877</td>\n",
       "      <td>-3.262881</td>\n",
       "      <td>-2.760550</td>\n",
       "      <td>1.420907</td>\n",
       "      <td>2.284498</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.036261</td>\n",
       "      <td>0.071000</td>\n",
       "      <td>-0.016092</td>\n",
       "      <td>0.081974</td>\n",
       "      <td>0.034969</td>\n",
       "      <td>0.061226</td>\n",
       "      <td>-0.012242</td>\n",
       "      <td>-0.056969</td>\n",
       "      <td>0.023906</td>\n",
       "      <td>-8.201801e-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1016 rows × 665 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene expression__Feature_exp_pca0  \\\n",
       "PUBLIC_ID                                      \n",
       "MMRF_1007                                NaN   \n",
       "MMRF_1011                                NaN   \n",
       "MMRF_1013                                NaN   \n",
       "MMRF_1014                                NaN   \n",
       "MMRF_1016                                NaN   \n",
       "...                                      ...   \n",
       "MMRF_2846                                NaN   \n",
       "MMRF_2847                           0.115155   \n",
       "MMRF_2848                          -4.057280   \n",
       "MMRF_2851                          -3.540592   \n",
       "MMRF_2853                           1.753795   \n",
       "\n",
       "           Gene expression__Feature_exp_pca1  \\\n",
       "PUBLIC_ID                                      \n",
       "MMRF_1007                                NaN   \n",
       "MMRF_1011                                NaN   \n",
       "MMRF_1013                                NaN   \n",
       "MMRF_1014                                NaN   \n",
       "MMRF_1016                                NaN   \n",
       "...                                      ...   \n",
       "MMRF_2846                                NaN   \n",
       "MMRF_2847                          -0.262712   \n",
       "MMRF_2848                           5.160718   \n",
       "MMRF_2851                          -1.420694   \n",
       "MMRF_2853                          -0.998007   \n",
       "\n",
       "           Gene expression__Feature_exp_pca2  \\\n",
       "PUBLIC_ID                                      \n",
       "MMRF_1007                                NaN   \n",
       "MMRF_1011                                NaN   \n",
       "MMRF_1013                                NaN   \n",
       "MMRF_1014                                NaN   \n",
       "MMRF_1016                                NaN   \n",
       "...                                      ...   \n",
       "MMRF_2846                                NaN   \n",
       "MMRF_2847                           9.547609   \n",
       "MMRF_2848                           7.087989   \n",
       "MMRF_2851                           6.074475   \n",
       "MMRF_2853                           3.684521   \n",
       "\n",
       "           Gene expression__Feature_exp_pca3  \\\n",
       "PUBLIC_ID                                      \n",
       "MMRF_1007                                NaN   \n",
       "MMRF_1011                                NaN   \n",
       "MMRF_1013                                NaN   \n",
       "MMRF_1014                                NaN   \n",
       "MMRF_1016                                NaN   \n",
       "...                                      ...   \n",
       "MMRF_2846                                NaN   \n",
       "MMRF_2847                           0.174491   \n",
       "MMRF_2848                          -0.509473   \n",
       "MMRF_2851                           1.946850   \n",
       "MMRF_2853                          -4.461104   \n",
       "\n",
       "           Gene expression__Feature_exp_pca4  \\\n",
       "PUBLIC_ID                                      \n",
       "MMRF_1007                                NaN   \n",
       "MMRF_1011                                NaN   \n",
       "MMRF_1013                                NaN   \n",
       "MMRF_1014                                NaN   \n",
       "MMRF_1016                                NaN   \n",
       "...                                      ...   \n",
       "MMRF_2846                                NaN   \n",
       "MMRF_2847                          -1.700913   \n",
       "MMRF_2848                           0.181885   \n",
       "MMRF_2851                           2.362985   \n",
       "MMRF_2853                          -7.864048   \n",
       "\n",
       "           Gene expression__Feature_exp_pca5  \\\n",
       "PUBLIC_ID                                      \n",
       "MMRF_1007                                NaN   \n",
       "MMRF_1011                                NaN   \n",
       "MMRF_1013                                NaN   \n",
       "MMRF_1014                                NaN   \n",
       "MMRF_1016                                NaN   \n",
       "...                                      ...   \n",
       "MMRF_2846                                NaN   \n",
       "MMRF_2847                          -2.594641   \n",
       "MMRF_2848                           2.476096   \n",
       "MMRF_2851                          -1.068620   \n",
       "MMRF_2853                           0.810877   \n",
       "\n",
       "           Gene expression__Feature_exp_pca6  \\\n",
       "PUBLIC_ID                                      \n",
       "MMRF_1007                                NaN   \n",
       "MMRF_1011                                NaN   \n",
       "MMRF_1013                                NaN   \n",
       "MMRF_1014                                NaN   \n",
       "MMRF_1016                                NaN   \n",
       "...                                      ...   \n",
       "MMRF_2846                                NaN   \n",
       "MMRF_2847                           0.453928   \n",
       "MMRF_2848                          -2.930544   \n",
       "MMRF_2851                          -0.522922   \n",
       "MMRF_2853                          -3.262881   \n",
       "\n",
       "           Gene expression__Feature_exp_pca7  \\\n",
       "PUBLIC_ID                                      \n",
       "MMRF_1007                                NaN   \n",
       "MMRF_1011                                NaN   \n",
       "MMRF_1013                                NaN   \n",
       "MMRF_1014                                NaN   \n",
       "MMRF_1016                                NaN   \n",
       "...                                      ...   \n",
       "MMRF_2846                                NaN   \n",
       "MMRF_2847                           2.332060   \n",
       "MMRF_2848                           2.171186   \n",
       "MMRF_2851                          -0.722730   \n",
       "MMRF_2853                          -2.760550   \n",
       "\n",
       "           Gene expression__Feature_exp_pca8  \\\n",
       "PUBLIC_ID                                      \n",
       "MMRF_1007                                NaN   \n",
       "MMRF_1011                                NaN   \n",
       "MMRF_1013                                NaN   \n",
       "MMRF_1014                                NaN   \n",
       "MMRF_1016                                NaN   \n",
       "...                                      ...   \n",
       "MMRF_2846                                NaN   \n",
       "MMRF_2847                           3.599387   \n",
       "MMRF_2848                          -0.907540   \n",
       "MMRF_2851                          -1.051358   \n",
       "MMRF_2853                           1.420907   \n",
       "\n",
       "           Gene expression__Feature_exp_pca9  ...  \\\n",
       "PUBLIC_ID                                     ...   \n",
       "MMRF_1007                                NaN  ...   \n",
       "MMRF_1011                                NaN  ...   \n",
       "MMRF_1013                                NaN  ...   \n",
       "MMRF_1014                                NaN  ...   \n",
       "MMRF_1016                                NaN  ...   \n",
       "...                                      ...  ...   \n",
       "MMRF_2846                                NaN  ...   \n",
       "MMRF_2847                           3.843321  ...   \n",
       "MMRF_2848                           1.734954  ...   \n",
       "MMRF_2851                          -1.105945  ...   \n",
       "MMRF_2853                           2.284498  ...   \n",
       "\n",
       "           Gene expression__Feature_exp_pca655  \\\n",
       "PUBLIC_ID                                        \n",
       "MMRF_1007                                  NaN   \n",
       "MMRF_1011                                  NaN   \n",
       "MMRF_1013                                  NaN   \n",
       "MMRF_1014                                  NaN   \n",
       "MMRF_1016                                  NaN   \n",
       "...                                        ...   \n",
       "MMRF_2846                                  NaN   \n",
       "MMRF_2847                             0.001026   \n",
       "MMRF_2848                            -0.010497   \n",
       "MMRF_2851                            -0.010703   \n",
       "MMRF_2853                            -0.036261   \n",
       "\n",
       "           Gene expression__Feature_exp_pca656  \\\n",
       "PUBLIC_ID                                        \n",
       "MMRF_1007                                  NaN   \n",
       "MMRF_1011                                  NaN   \n",
       "MMRF_1013                                  NaN   \n",
       "MMRF_1014                                  NaN   \n",
       "MMRF_1016                                  NaN   \n",
       "...                                        ...   \n",
       "MMRF_2846                                  NaN   \n",
       "MMRF_2847                             0.007154   \n",
       "MMRF_2848                            -0.012944   \n",
       "MMRF_2851                            -0.005334   \n",
       "MMRF_2853                             0.071000   \n",
       "\n",
       "           Gene expression__Feature_exp_pca657  \\\n",
       "PUBLIC_ID                                        \n",
       "MMRF_1007                                  NaN   \n",
       "MMRF_1011                                  NaN   \n",
       "MMRF_1013                                  NaN   \n",
       "MMRF_1014                                  NaN   \n",
       "MMRF_1016                                  NaN   \n",
       "...                                        ...   \n",
       "MMRF_2846                                  NaN   \n",
       "MMRF_2847                            -0.035740   \n",
       "MMRF_2848                             0.043653   \n",
       "MMRF_2851                             0.057544   \n",
       "MMRF_2853                            -0.016092   \n",
       "\n",
       "           Gene expression__Feature_exp_pca658  \\\n",
       "PUBLIC_ID                                        \n",
       "MMRF_1007                                  NaN   \n",
       "MMRF_1011                                  NaN   \n",
       "MMRF_1013                                  NaN   \n",
       "MMRF_1014                                  NaN   \n",
       "MMRF_1016                                  NaN   \n",
       "...                                        ...   \n",
       "MMRF_2846                                  NaN   \n",
       "MMRF_2847                            -0.027436   \n",
       "MMRF_2848                             0.027963   \n",
       "MMRF_2851                             0.053831   \n",
       "MMRF_2853                             0.081974   \n",
       "\n",
       "           Gene expression__Feature_exp_pca659  \\\n",
       "PUBLIC_ID                                        \n",
       "MMRF_1007                                  NaN   \n",
       "MMRF_1011                                  NaN   \n",
       "MMRF_1013                                  NaN   \n",
       "MMRF_1014                                  NaN   \n",
       "MMRF_1016                                  NaN   \n",
       "...                                        ...   \n",
       "MMRF_2846                                  NaN   \n",
       "MMRF_2847                             0.005234   \n",
       "MMRF_2848                             0.014943   \n",
       "MMRF_2851                             0.067620   \n",
       "MMRF_2853                             0.034969   \n",
       "\n",
       "           Gene expression__Feature_exp_pca660  \\\n",
       "PUBLIC_ID                                        \n",
       "MMRF_1007                                  NaN   \n",
       "MMRF_1011                                  NaN   \n",
       "MMRF_1013                                  NaN   \n",
       "MMRF_1014                                  NaN   \n",
       "MMRF_1016                                  NaN   \n",
       "...                                        ...   \n",
       "MMRF_2846                                  NaN   \n",
       "MMRF_2847                             0.037007   \n",
       "MMRF_2848                             0.000232   \n",
       "MMRF_2851                             0.005861   \n",
       "MMRF_2853                             0.061226   \n",
       "\n",
       "           Gene expression__Feature_exp_pca661  \\\n",
       "PUBLIC_ID                                        \n",
       "MMRF_1007                                  NaN   \n",
       "MMRF_1011                                  NaN   \n",
       "MMRF_1013                                  NaN   \n",
       "MMRF_1014                                  NaN   \n",
       "MMRF_1016                                  NaN   \n",
       "...                                        ...   \n",
       "MMRF_2846                                  NaN   \n",
       "MMRF_2847                            -0.010941   \n",
       "MMRF_2848                            -0.031268   \n",
       "MMRF_2851                            -0.074586   \n",
       "MMRF_2853                            -0.012242   \n",
       "\n",
       "           Gene expression__Feature_exp_pca662  \\\n",
       "PUBLIC_ID                                        \n",
       "MMRF_1007                                  NaN   \n",
       "MMRF_1011                                  NaN   \n",
       "MMRF_1013                                  NaN   \n",
       "MMRF_1014                                  NaN   \n",
       "MMRF_1016                                  NaN   \n",
       "...                                        ...   \n",
       "MMRF_2846                                  NaN   \n",
       "MMRF_2847                             0.129442   \n",
       "MMRF_2848                             0.041034   \n",
       "MMRF_2851                             0.006660   \n",
       "MMRF_2853                            -0.056969   \n",
       "\n",
       "           Gene expression__Feature_exp_pca663  \\\n",
       "PUBLIC_ID                                        \n",
       "MMRF_1007                                  NaN   \n",
       "MMRF_1011                                  NaN   \n",
       "MMRF_1013                                  NaN   \n",
       "MMRF_1014                                  NaN   \n",
       "MMRF_1016                                  NaN   \n",
       "...                                        ...   \n",
       "MMRF_2846                                  NaN   \n",
       "MMRF_2847                            -0.037312   \n",
       "MMRF_2848                             0.000061   \n",
       "MMRF_2851                            -0.009274   \n",
       "MMRF_2853                             0.023906   \n",
       "\n",
       "           Gene expression__Feature_exp_pca664  \n",
       "PUBLIC_ID                                       \n",
       "MMRF_1007                                  NaN  \n",
       "MMRF_1011                                  NaN  \n",
       "MMRF_1013                                  NaN  \n",
       "MMRF_1014                                  NaN  \n",
       "MMRF_1016                                  NaN  \n",
       "...                                        ...  \n",
       "MMRF_2846                                  NaN  \n",
       "MMRF_2847                         1.899866e-15  \n",
       "MMRF_2848                         7.341321e-16  \n",
       "MMRF_2851                        -6.120133e-16  \n",
       "MMRF_2853                        -8.201801e-16  \n",
       "\n",
       "[1016 rows x 665 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out.filter(regex='Feature_exp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2373564",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca0</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca1</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca2</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca3</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca4</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca5</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca6</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca7</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca8</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca9</th>\n",
       "      <th>...</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca87</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca88</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca89</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca90</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca91</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca92</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca93</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca94</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca95</th>\n",
       "      <th>Gistic copy number__Feature_CN_gisticpca96</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PUBLIC_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MMRF_1007</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_1011</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_1013</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_1014</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_1016</th>\n",
       "      <td>2.490232</td>\n",
       "      <td>2.648216</td>\n",
       "      <td>-2.598980</td>\n",
       "      <td>0.724026</td>\n",
       "      <td>1.147801</td>\n",
       "      <td>-0.495525</td>\n",
       "      <td>-1.797174</td>\n",
       "      <td>0.024392</td>\n",
       "      <td>-0.472550</td>\n",
       "      <td>-0.614534</td>\n",
       "      <td>...</td>\n",
       "      <td>0.421373</td>\n",
       "      <td>-0.158997</td>\n",
       "      <td>-0.106580</td>\n",
       "      <td>0.130419</td>\n",
       "      <td>0.048369</td>\n",
       "      <td>0.054240</td>\n",
       "      <td>0.036508</td>\n",
       "      <td>-0.003299</td>\n",
       "      <td>-0.010472</td>\n",
       "      <td>-0.033015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_2846</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_2847</th>\n",
       "      <td>1.365294</td>\n",
       "      <td>-1.561895</td>\n",
       "      <td>0.678246</td>\n",
       "      <td>0.228827</td>\n",
       "      <td>0.070804</td>\n",
       "      <td>-0.536736</td>\n",
       "      <td>-0.147391</td>\n",
       "      <td>1.997537</td>\n",
       "      <td>-0.253596</td>\n",
       "      <td>-0.849476</td>\n",
       "      <td>...</td>\n",
       "      <td>0.563214</td>\n",
       "      <td>-0.126622</td>\n",
       "      <td>-0.056641</td>\n",
       "      <td>-0.121586</td>\n",
       "      <td>0.091660</td>\n",
       "      <td>-0.116242</td>\n",
       "      <td>0.021429</td>\n",
       "      <td>0.018906</td>\n",
       "      <td>0.026588</td>\n",
       "      <td>-0.041085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_2848</th>\n",
       "      <td>-1.133501</td>\n",
       "      <td>2.237886</td>\n",
       "      <td>-0.598636</td>\n",
       "      <td>0.017751</td>\n",
       "      <td>0.983006</td>\n",
       "      <td>0.713377</td>\n",
       "      <td>-0.762593</td>\n",
       "      <td>-0.244717</td>\n",
       "      <td>-0.244440</td>\n",
       "      <td>-0.688495</td>\n",
       "      <td>...</td>\n",
       "      <td>0.100512</td>\n",
       "      <td>0.359694</td>\n",
       "      <td>0.006428</td>\n",
       "      <td>-0.147517</td>\n",
       "      <td>0.000310</td>\n",
       "      <td>-0.015194</td>\n",
       "      <td>-0.091654</td>\n",
       "      <td>-0.030962</td>\n",
       "      <td>-0.071680</td>\n",
       "      <td>-0.019823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_2851</th>\n",
       "      <td>-2.924364</td>\n",
       "      <td>-0.254533</td>\n",
       "      <td>1.515942</td>\n",
       "      <td>0.182573</td>\n",
       "      <td>-0.457697</td>\n",
       "      <td>-0.215411</td>\n",
       "      <td>-0.218937</td>\n",
       "      <td>-0.475667</td>\n",
       "      <td>-0.608096</td>\n",
       "      <td>0.033646</td>\n",
       "      <td>...</td>\n",
       "      <td>0.093671</td>\n",
       "      <td>-0.188176</td>\n",
       "      <td>0.084344</td>\n",
       "      <td>-0.074361</td>\n",
       "      <td>-0.060737</td>\n",
       "      <td>-0.346298</td>\n",
       "      <td>0.295009</td>\n",
       "      <td>-0.267096</td>\n",
       "      <td>-0.033595</td>\n",
       "      <td>-0.033732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMRF_2853</th>\n",
       "      <td>1.713711</td>\n",
       "      <td>0.823067</td>\n",
       "      <td>1.320051</td>\n",
       "      <td>-2.146064</td>\n",
       "      <td>3.010541</td>\n",
       "      <td>-0.198010</td>\n",
       "      <td>-0.983387</td>\n",
       "      <td>-0.457649</td>\n",
       "      <td>1.761399</td>\n",
       "      <td>0.355651</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.113064</td>\n",
       "      <td>-0.056066</td>\n",
       "      <td>0.060814</td>\n",
       "      <td>-0.344480</td>\n",
       "      <td>-0.328171</td>\n",
       "      <td>-0.077196</td>\n",
       "      <td>-0.071439</td>\n",
       "      <td>0.003473</td>\n",
       "      <td>0.037953</td>\n",
       "      <td>-0.090142</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1016 rows × 97 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gistic copy number__Feature_CN_gisticpca0  \\\n",
       "PUBLIC_ID                                              \n",
       "MMRF_1007                                        NaN   \n",
       "MMRF_1011                                        NaN   \n",
       "MMRF_1013                                        NaN   \n",
       "MMRF_1014                                        NaN   \n",
       "MMRF_1016                                   2.490232   \n",
       "...                                              ...   \n",
       "MMRF_2846                                        NaN   \n",
       "MMRF_2847                                   1.365294   \n",
       "MMRF_2848                                  -1.133501   \n",
       "MMRF_2851                                  -2.924364   \n",
       "MMRF_2853                                   1.713711   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca1  \\\n",
       "PUBLIC_ID                                              \n",
       "MMRF_1007                                        NaN   \n",
       "MMRF_1011                                        NaN   \n",
       "MMRF_1013                                        NaN   \n",
       "MMRF_1014                                        NaN   \n",
       "MMRF_1016                                   2.648216   \n",
       "...                                              ...   \n",
       "MMRF_2846                                        NaN   \n",
       "MMRF_2847                                  -1.561895   \n",
       "MMRF_2848                                   2.237886   \n",
       "MMRF_2851                                  -0.254533   \n",
       "MMRF_2853                                   0.823067   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca2  \\\n",
       "PUBLIC_ID                                              \n",
       "MMRF_1007                                        NaN   \n",
       "MMRF_1011                                        NaN   \n",
       "MMRF_1013                                        NaN   \n",
       "MMRF_1014                                        NaN   \n",
       "MMRF_1016                                  -2.598980   \n",
       "...                                              ...   \n",
       "MMRF_2846                                        NaN   \n",
       "MMRF_2847                                   0.678246   \n",
       "MMRF_2848                                  -0.598636   \n",
       "MMRF_2851                                   1.515942   \n",
       "MMRF_2853                                   1.320051   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca3  \\\n",
       "PUBLIC_ID                                              \n",
       "MMRF_1007                                        NaN   \n",
       "MMRF_1011                                        NaN   \n",
       "MMRF_1013                                        NaN   \n",
       "MMRF_1014                                        NaN   \n",
       "MMRF_1016                                   0.724026   \n",
       "...                                              ...   \n",
       "MMRF_2846                                        NaN   \n",
       "MMRF_2847                                   0.228827   \n",
       "MMRF_2848                                   0.017751   \n",
       "MMRF_2851                                   0.182573   \n",
       "MMRF_2853                                  -2.146064   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca4  \\\n",
       "PUBLIC_ID                                              \n",
       "MMRF_1007                                        NaN   \n",
       "MMRF_1011                                        NaN   \n",
       "MMRF_1013                                        NaN   \n",
       "MMRF_1014                                        NaN   \n",
       "MMRF_1016                                   1.147801   \n",
       "...                                              ...   \n",
       "MMRF_2846                                        NaN   \n",
       "MMRF_2847                                   0.070804   \n",
       "MMRF_2848                                   0.983006   \n",
       "MMRF_2851                                  -0.457697   \n",
       "MMRF_2853                                   3.010541   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca5  \\\n",
       "PUBLIC_ID                                              \n",
       "MMRF_1007                                        NaN   \n",
       "MMRF_1011                                        NaN   \n",
       "MMRF_1013                                        NaN   \n",
       "MMRF_1014                                        NaN   \n",
       "MMRF_1016                                  -0.495525   \n",
       "...                                              ...   \n",
       "MMRF_2846                                        NaN   \n",
       "MMRF_2847                                  -0.536736   \n",
       "MMRF_2848                                   0.713377   \n",
       "MMRF_2851                                  -0.215411   \n",
       "MMRF_2853                                  -0.198010   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca6  \\\n",
       "PUBLIC_ID                                              \n",
       "MMRF_1007                                        NaN   \n",
       "MMRF_1011                                        NaN   \n",
       "MMRF_1013                                        NaN   \n",
       "MMRF_1014                                        NaN   \n",
       "MMRF_1016                                  -1.797174   \n",
       "...                                              ...   \n",
       "MMRF_2846                                        NaN   \n",
       "MMRF_2847                                  -0.147391   \n",
       "MMRF_2848                                  -0.762593   \n",
       "MMRF_2851                                  -0.218937   \n",
       "MMRF_2853                                  -0.983387   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca7  \\\n",
       "PUBLIC_ID                                              \n",
       "MMRF_1007                                        NaN   \n",
       "MMRF_1011                                        NaN   \n",
       "MMRF_1013                                        NaN   \n",
       "MMRF_1014                                        NaN   \n",
       "MMRF_1016                                   0.024392   \n",
       "...                                              ...   \n",
       "MMRF_2846                                        NaN   \n",
       "MMRF_2847                                   1.997537   \n",
       "MMRF_2848                                  -0.244717   \n",
       "MMRF_2851                                  -0.475667   \n",
       "MMRF_2853                                  -0.457649   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca8  \\\n",
       "PUBLIC_ID                                              \n",
       "MMRF_1007                                        NaN   \n",
       "MMRF_1011                                        NaN   \n",
       "MMRF_1013                                        NaN   \n",
       "MMRF_1014                                        NaN   \n",
       "MMRF_1016                                  -0.472550   \n",
       "...                                              ...   \n",
       "MMRF_2846                                        NaN   \n",
       "MMRF_2847                                  -0.253596   \n",
       "MMRF_2848                                  -0.244440   \n",
       "MMRF_2851                                  -0.608096   \n",
       "MMRF_2853                                   1.761399   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca9  ...  \\\n",
       "PUBLIC_ID                                             ...   \n",
       "MMRF_1007                                        NaN  ...   \n",
       "MMRF_1011                                        NaN  ...   \n",
       "MMRF_1013                                        NaN  ...   \n",
       "MMRF_1014                                        NaN  ...   \n",
       "MMRF_1016                                  -0.614534  ...   \n",
       "...                                              ...  ...   \n",
       "MMRF_2846                                        NaN  ...   \n",
       "MMRF_2847                                  -0.849476  ...   \n",
       "MMRF_2848                                  -0.688495  ...   \n",
       "MMRF_2851                                   0.033646  ...   \n",
       "MMRF_2853                                   0.355651  ...   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca87  \\\n",
       "PUBLIC_ID                                               \n",
       "MMRF_1007                                         NaN   \n",
       "MMRF_1011                                         NaN   \n",
       "MMRF_1013                                         NaN   \n",
       "MMRF_1014                                         NaN   \n",
       "MMRF_1016                                    0.421373   \n",
       "...                                               ...   \n",
       "MMRF_2846                                         NaN   \n",
       "MMRF_2847                                    0.563214   \n",
       "MMRF_2848                                    0.100512   \n",
       "MMRF_2851                                    0.093671   \n",
       "MMRF_2853                                   -0.113064   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca88  \\\n",
       "PUBLIC_ID                                               \n",
       "MMRF_1007                                         NaN   \n",
       "MMRF_1011                                         NaN   \n",
       "MMRF_1013                                         NaN   \n",
       "MMRF_1014                                         NaN   \n",
       "MMRF_1016                                   -0.158997   \n",
       "...                                               ...   \n",
       "MMRF_2846                                         NaN   \n",
       "MMRF_2847                                   -0.126622   \n",
       "MMRF_2848                                    0.359694   \n",
       "MMRF_2851                                   -0.188176   \n",
       "MMRF_2853                                   -0.056066   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca89  \\\n",
       "PUBLIC_ID                                               \n",
       "MMRF_1007                                         NaN   \n",
       "MMRF_1011                                         NaN   \n",
       "MMRF_1013                                         NaN   \n",
       "MMRF_1014                                         NaN   \n",
       "MMRF_1016                                   -0.106580   \n",
       "...                                               ...   \n",
       "MMRF_2846                                         NaN   \n",
       "MMRF_2847                                   -0.056641   \n",
       "MMRF_2848                                    0.006428   \n",
       "MMRF_2851                                    0.084344   \n",
       "MMRF_2853                                    0.060814   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca90  \\\n",
       "PUBLIC_ID                                               \n",
       "MMRF_1007                                         NaN   \n",
       "MMRF_1011                                         NaN   \n",
       "MMRF_1013                                         NaN   \n",
       "MMRF_1014                                         NaN   \n",
       "MMRF_1016                                    0.130419   \n",
       "...                                               ...   \n",
       "MMRF_2846                                         NaN   \n",
       "MMRF_2847                                   -0.121586   \n",
       "MMRF_2848                                   -0.147517   \n",
       "MMRF_2851                                   -0.074361   \n",
       "MMRF_2853                                   -0.344480   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca91  \\\n",
       "PUBLIC_ID                                               \n",
       "MMRF_1007                                         NaN   \n",
       "MMRF_1011                                         NaN   \n",
       "MMRF_1013                                         NaN   \n",
       "MMRF_1014                                         NaN   \n",
       "MMRF_1016                                    0.048369   \n",
       "...                                               ...   \n",
       "MMRF_2846                                         NaN   \n",
       "MMRF_2847                                    0.091660   \n",
       "MMRF_2848                                    0.000310   \n",
       "MMRF_2851                                   -0.060737   \n",
       "MMRF_2853                                   -0.328171   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca92  \\\n",
       "PUBLIC_ID                                               \n",
       "MMRF_1007                                         NaN   \n",
       "MMRF_1011                                         NaN   \n",
       "MMRF_1013                                         NaN   \n",
       "MMRF_1014                                         NaN   \n",
       "MMRF_1016                                    0.054240   \n",
       "...                                               ...   \n",
       "MMRF_2846                                         NaN   \n",
       "MMRF_2847                                   -0.116242   \n",
       "MMRF_2848                                   -0.015194   \n",
       "MMRF_2851                                   -0.346298   \n",
       "MMRF_2853                                   -0.077196   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca93  \\\n",
       "PUBLIC_ID                                               \n",
       "MMRF_1007                                         NaN   \n",
       "MMRF_1011                                         NaN   \n",
       "MMRF_1013                                         NaN   \n",
       "MMRF_1014                                         NaN   \n",
       "MMRF_1016                                    0.036508   \n",
       "...                                               ...   \n",
       "MMRF_2846                                         NaN   \n",
       "MMRF_2847                                    0.021429   \n",
       "MMRF_2848                                   -0.091654   \n",
       "MMRF_2851                                    0.295009   \n",
       "MMRF_2853                                   -0.071439   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca94  \\\n",
       "PUBLIC_ID                                               \n",
       "MMRF_1007                                         NaN   \n",
       "MMRF_1011                                         NaN   \n",
       "MMRF_1013                                         NaN   \n",
       "MMRF_1014                                         NaN   \n",
       "MMRF_1016                                   -0.003299   \n",
       "...                                               ...   \n",
       "MMRF_2846                                         NaN   \n",
       "MMRF_2847                                    0.018906   \n",
       "MMRF_2848                                   -0.030962   \n",
       "MMRF_2851                                   -0.267096   \n",
       "MMRF_2853                                    0.003473   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca95  \\\n",
       "PUBLIC_ID                                               \n",
       "MMRF_1007                                         NaN   \n",
       "MMRF_1011                                         NaN   \n",
       "MMRF_1013                                         NaN   \n",
       "MMRF_1014                                         NaN   \n",
       "MMRF_1016                                   -0.010472   \n",
       "...                                               ...   \n",
       "MMRF_2846                                         NaN   \n",
       "MMRF_2847                                    0.026588   \n",
       "MMRF_2848                                   -0.071680   \n",
       "MMRF_2851                                   -0.033595   \n",
       "MMRF_2853                                    0.037953   \n",
       "\n",
       "           Gistic copy number__Feature_CN_gisticpca96  \n",
       "PUBLIC_ID                                              \n",
       "MMRF_1007                                         NaN  \n",
       "MMRF_1011                                         NaN  \n",
       "MMRF_1013                                         NaN  \n",
       "MMRF_1014                                         NaN  \n",
       "MMRF_1016                                   -0.033015  \n",
       "...                                               ...  \n",
       "MMRF_2846                                         NaN  \n",
       "MMRF_2847                                   -0.041085  \n",
       "MMRF_2848                                   -0.019823  \n",
       "MMRF_2851                                   -0.033732  \n",
       "MMRF_2853                                   -0.090142  \n",
       "\n",
       "[1016 rows x 97 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "out.filter(regex='Feature_CN_gistic')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecf63b67-8454-4b85-b72d-4b01aba4258c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "out = pipeline.fit_transform(features, y)\n",
    "print(out.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c247d94-5c05-4532-8e34-2dbcec5d1c4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "out.filter(regex='Gistic copy number').shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52424d59-918a-46c0-99c1-2917af19a683",
   "metadata": {},
   "outputs": [],
   "source": [
    "out.filter(regex='Gene copy number').shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "290b675a",
   "metadata": {},
   "outputs": [],
   "source": [
    "out.filter(regex='Gene expression').shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14a86fb5-1fed-4207-8e9e-552e4fdf2152",
   "metadata": {},
   "outputs": [],
   "source": [
    "out.filter(regex='FISH copy number').shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a924f6a9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72dab04f",
   "metadata": {},
   "outputs": [],
   "source": [
    "scratchdir=\"/scratch/users/nus/e1083772/cancer-survival-ml/data/splits\"\n",
    "\n",
    "train_out_features_file=f'{scratchdir}/{shuffle}/{fold}/features_subset.parquet'\n",
    "valid_out_features_file=f'{scratchdir}/{shuffle}/{fold}/valid_features_subset.parquet'\n",
    "\n",
    "out.to_parquet(train_out_features_file)\n",
    "outv.to_parquet(valid_out_features_file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
